File: src/Screen.java
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import static java.lang.Math.min;
import java.util.Arrays;
import java.util.LinkedList;
import javax.imageio.ImageIO;
import org.tukaani.xz.SeekableFileInputStream;
import org.tukaani.xz.SeekableXZInputStream;

public class Screen extends Main {
  private static final int FRAME_SIZE = BYTE_SIZE + ATTR_SIZE;
  private static final int[] backgroundOn = new int[PIXEL_SIZE];
  private static int[] attrs;

  private static SeekableXZInputStream in;
  
  public static void init() throws IOException {
    SeekableFileInputStream file
        = new SeekableFileInputStream(project + "video.xz");
    in = new SeekableXZInputStream(file);
    loadBackgrounds();
  }
  
  // loading
  
  private static boolean[] load(int num) throws IOException {
    boolean[] data = new boolean[PIXEL_SIZE];
    
    byte[] byteScreen = new byte[FRAME_SIZE];
    in.seek(num * FRAME_SIZE);
    if(in.read(byteScreen, 0, FRAME_SIZE) < FRAME_SIZE) throw new IOException();

    attrs = new int[ATTR_SIZE];
    for(int x = 0; x < ATTR_SIZE; x++) {
      int value = byteScreen[BYTE_SIZE | x];
      int bright = (value & 0b1000000) == 0 ? 0 : 0b10001000;
      attrs[x] =  (value & 0b111) | ((value & 0b111000) << 1) | bright;
    }
    
    main: for(int part = 0; part < 3; part++) {
      int partSource = part << 11;
      int partDestination = part << 3;
      for(int y = 0; y < 8; y++) {
        int ySource = partSource | (y << 5);
        int yDestination = partDestination | y;
        if(yDestination < AREA_Y) continue;
        if(yDestination >= MAX_AREA_Y) break main;
        yDestination = (yDestination - AREA_Y) << 3;
        for(int yy = 0; yy < 8; yy++) {
          int yySource = ySource | (yy << 8);
          int yyDestination = (yDestination | yy) * AREA_WIDTH;
          for(int x = 0; x < AREA_WIDTH; x++) {
            int source = yySource | (x + AREA_X);
            int destination = (yyDestination + x) << 3;
            int byteValue = byteScreen[source];
            for(int xx = 7; xx >= 0; xx--) {
              int addr = destination | xx;
              boolean pixelValue = (byteValue & 1) > 0;
              data[addr] = pixelValue;
              byteValue = byteValue >> 1;
            }
          }
        }
      }
    }
    
    return data;
  }
  
  // background
  
  private static class Background {
    private final boolean[] values;
    private final BufferedImage image;
    private final String fileName;

    public Background(boolean[] values) {
      this.values = values;
      this.image = null;
      this.fileName = "";
    }

    public Background(boolean[] values, BufferedImage image, String fileName) {
      this.values = values;
      this.image = image;
      this.fileName = fileName;
    }
    
    public int difference(boolean[] screen) {
      int difference = 0;
      for(int i = 0; i < PIXEL_SIZE; i++)
        if(values[i] != screen[i]) difference++;
      return difference;
    }
  }
  
  public static void loadBackgrounds() throws IOException {
    for(final File file: new File(project + "backgrounds").listFiles()) {
      if(file.isDirectory()) continue;
      BufferedImage image = ImageIO.read(file);
      boolean[] pixels = new boolean[PIXEL_SIZE];
      for(int y = 0; y < PIXEL_HEIGHT; y++) {
        for(int x = 0; x < PIXEL_WIDTH; x++) {
          pixels[x + y * PIXEL_WIDTH] = (image.getRGB(x, y) & 0xFF) > 0x7F;
        }
      }
      backgrounds.add(new Background(pixels, null, file.getName()));
    }
  }
  
  public static void saveBackgrounds() throws IOException {
    for(Background background: backgrounds) {
      int number = Integer.parseInt(
          background.fileName.substring(0, background.fileName.indexOf('.')));
      load(number + 1);
      saveImage(toImage(background.values, null), background.fileName);
    }
  }
  
  private static Background findBackground(boolean[] screen) {
    int minDifference = PIXEL_SIZE;
    Background minBackground = null;
    for(Background background: backgrounds) {
      int difference = background.difference(screen);
      if(difference < minDifference) {
        minDifference = difference;
        minBackground = background;
      }
    }
    //System.out.println("Min difference is " + minDifference);
    return minBackground;
  }
  
  private static LinkedList<Background> backgrounds
      = new LinkedList<>();

  private static boolean[] composeBackground(int frames) {
    int minFrames = (int) Math.floor(PERCENT_ON * frames);
    boolean[] background = new boolean[PIXEL_SIZE];
    for(int addr = 0; addr < PIXEL_SIZE; addr++) 
      background[addr] = backgroundOn[addr] >= minFrames;
    return background;
  }
  
  // processing

  public static void process(int... start) throws IOException {
    for(int i = 0; i < start.length; i++) process(start[i], -1, true);
  }
  
  public static void process(int from, int to, boolean singleScreen)
      throws IOException {
    int firstFrame = from;
    boolean[] oldScreen = null;
    for(int frame = from; frame <= to || to < 0; frame++) {
      boolean[] screen;
      try {
        screen = load(frame);
      } catch (IOException ex) {
        break;
      }
      
      if(oldScreen == null) {
        oldScreen = screen;
        continue;
      }
      
      //final int SAME = 0, CHANGED = 1;
      //int[] pixels = new int[PIXEL_SIZE];
      int difference = 0;
      for(int addr = 0; addr < PIXEL_SIZE; addr++) {
        boolean isChanged = screen[addr] != oldScreen[addr];
        //pixels[addr] = isChanged ? CHANGED : SAME;
        if(screen[addr]) backgroundOn[addr]++;
        if(isChanged) difference++;
      }
        
      if(difference >= MIN_DIFFERENCE) {
        int frames = frame - firstFrame;
        if(frames >= MIN_FRAMES) {
          oldScreen = composeBackground(frames);
          if(mode == Mode.EXTRACT_BACKGROUNDS) {
            if(findBackground(oldScreen) == null || SAVE_SIMILAR) {
              saveImage(toImage(oldScreen, null), firstFrame);
              backgrounds.add(new Background(oldScreen));
              //saveImage(toImage(oldScreen, null), frame - 1);
              //saveImage(toImage(screen, null), frame);
              System.out.println("Saved background " + firstFrame
                  + " with difference " + difference + " and " + frames
                  + " frames");
            }
          } else {
            System.out.println("Sequence " + firstFrame + " - " + (frame - 1));
          }
        }
        if(singleScreen) return;
        Arrays.fill(backgroundOn, 0);
        firstFrame = frame;
      } else if(frame % FRAME_FREQUENCY == 0) {
        switch(mode) {
          case SHOW_DIFFERENCE:
            Background background = findBackground(screen);
            if(background != null) saveImage(toImage(screen, background.values)
                , frame);
            break;
          case TO_BLACK_AND_WHITE:
            saveImage(toImage(screen, null), frame);
            break;
        }
      }

      oldScreen = screen;
    }
  }
  
  // conversion and saving
  
  public static BufferedImage toImage(boolean[] screenData, boolean[] bgData) {
    BufferedImage image = new BufferedImage(PIXEL_WIDTH, PIXEL_HEIGHT
        , BufferedImage.TYPE_INT_RGB);
    
    for(int y = 0; y < PIXEL_HEIGHT; y++) {
      int ySource = y << 8;
      int yAttrSource = ((y >> 3) + AREA_Y) << 5;
      for(int x = 0; x < PIXEL_WIDTH; x++) {
        int attr = attrs[yAttrSource | (x >> 3 + AREA_X)];
        int addr = ySource | x;
        boolean value = screenData[addr];
        int col;
        if(BLACK_AND_WHITE) {
          col = color[value ? 15 : 0];
        } else {
          col = value ? color[attr & 15] : color[(attr >> 4) & 15];
        }
        if(bgData != null && value != bgData[addr]) col = color[11];
        image.setRGB(x, y, col);
      }
    }
    
    return image;
  }
  
  public static void saveImage(BufferedImage image, int fileNumber)
      throws IOException {
    saveImage(image, String.format("%06d", fileNumber) + ".png");
  }
  
  public static void saveImage(BufferedImage image, String fileName)
      throws IOException {
    File outputfile = new File(OUT_DIR + fileName);
    ImageIO.write(x3(image), "png", outputfile);
  }
}

File: src/Screen.kt
import Background.Companion.backgrounds
import Background.Companion.composeBackground
import Background.Companion.findBackground
import org.tukaani.xz.SeekableFileInputStream
import org.tukaani.xz.SeekableXZInputStream
import java.awt.Color
import java.awt.FontMetrics
import java.awt.Graphics2D
import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import java.util.*
import javax.imageio.ImageIO

object Screen {
  private val backgroundOn = IntArray(PIXEL_SIZE)
  private var inStream: SeekableXZInputStream? = null

  @Throws(IOException::class)
  fun init() {
    val file = SeekableFileInputStream("$project/video.xz")
    inStream = SeekableXZInputStream(file)
    Background.loadBackgrounds()
  }

  // loading
  @Throws(IOException::class)
  private fun load(num:Int, area: Rect): Area {
    val data = Array<Pixel>(area.pixelSize()) {Pixel.OFF}
    val attrs = IntArray(area.size())
    val byteScreen = ByteArray(FRAME_SIZE)
    inStream!!.seek((num * FRAME_SIZE).toLong())
    if(inStream!!.read(byteScreen, 0, FRAME_SIZE) < FRAME_SIZE)
      throw IOException()
    val attrOffset = BYTE_SIZE + (area.y shl 5)
    for(x in 0 until area.size()) {
      val value = byteScreen[x + attrOffset].toInt()
      val bright = if(value and 0b1000000 == 0) 0 else 0b10001000
      attrs[x] = value and 0b111 or (value and 0b111000 shl 1) or bright
    }
    main@ for(part in 0..2) {
      val partSource = part shl 11
      val partDestination = part shl 3
      for(y in 0 until 8) {
        val ySource = partSource or (y shl 5)
        var yDestination = partDestination or y
        if(yDestination < area.y) continue
        if(yDestination >= area.y + area.height) break@main
        yDestination = yDestination - area.y shl 3
        for(yy in 0 until 8) {
          val yySource = ySource or (yy shl 8)
          val yyDestination = (yDestination or yy) * area.width
          for(x in 0 until area.width) {
            val source = yySource or x + area.x
            val destination = yyDestination + x shl 3
            var byteValue = byteScreen[source].toInt()
            for(xx in 7 downTo 0) {
              val addr = destination or xx
              val pixelValue = if(byteValue and 1 > 0) Pixel.ON else Pixel.OFF
              data[addr] = pixelValue
              byteValue = byteValue shr 1
            }
          }
        }
      }
    }
    return Area(data, attrs, area)
  }

  // processing

  fun process() {
    process(0, -1)
  }

  @Throws(IOException::class)
  fun process(from: Int, to: Int, frequency: Int = 1) {
    var firstFrame = from
    var oldScreen: Area? = null
    var frame = from
    while(frame <= to || to < 0) {
      if(frame % 1000 == 0) println(frame)

      val screen: Area = try {
        load(frame, MAIN_SCREEN)
      } catch(ex: IOException) {
        break
      }

      if(mode == Mode.SCREENSHOTS) {
        saveImage(toImage(load(frame, WHOLE_SCREEN), true), frame)
        frame++
        continue
      }

      if(oldScreen == null) {
        oldScreen = screen
        frame++
        continue
      }

      if(!ONLY_ABSENT || !File("D:\\output_final\\"
            + String.format("%06d", frame) + ".png").exists()) {
        if(mode == Mode.COLOR_BACKGROUNDS) {
          val background = Background.findBackground(screen.pixels, frame, false)
          if(background != null && background.frame < 0) {
            background.frame = frame
          }
        } else if(mode == Mode.EXTRACT_BACKGROUNDS) {
          println("$frame background switch")

          var difference = 0
          for(addr in 0 until PIXEL_SIZE) {
            val isChanged = screen.pixels[addr] != oldScreen.pixels[addr]
            if(screen.pixels[addr] == Pixel.ON) backgroundOn[addr]++
            if(isChanged) difference++
          }

          if(difference > MAX_BG_DIFFERENCE) {
            println("Processing sequence $firstFrame - $frame "
                  + if(mode === Mode.EXTRACT_SPRITES) ", "
                  + ImageExtractor.images.size else ""
            )
            val frames = frame - firstFrame
            if(frames >= MIN_FRAMES) {
              val background = composeBackground(frames, backgroundOn)
              if(SAVE_SIMILAR || findBackground(background, frame
                  , false) == null) {
                oldScreen = Area(background, oldScreen.attrs, oldScreen.area)
                saveImage(toImage(oldScreen, false), firstFrame)
                Background.addBackground(background)
                //saveImage(toImage(oldScreen, null), frame - 1)
                //saveImage(toImage(screen, null), frame)
                println("Saved background $firstFrame with difference"
                    + " $difference and $frames frames")
              }
            }
            Arrays.fill(backgroundOn, 0)
            firstFrame = frame
          }
        } else if(mode == Mode.FIND_PIXELS_TO_SKIP) {
          val background = findBackground(screen.pixels, frame
            , ONLY_BACKGROUND.isNotEmpty())
          if(background != null) {
            background.frame = frame
            background.total++
            for(addr in 0 until PIXEL_SIZE) {
              if(screen.pixels[addr] != background.pixels[addr]) {
                background.changed!![addr]++
              }
            }
          }
        } else if(frame % frequency == 0) {
          val background = findBackground(screen.pixels, frame
            , ONLY_BACKGROUND.isNotEmpty())
          if(background != null) {
            if(mode == Mode.SHOW_DIFFERENCE) {
              saveImage(toImage(screen, true, background.pixels), frame)
            } else {
              val image = if(background.image == null) {
                toImage(screen, true)
              } else {
                copyImage(background.image)
              }
              ImageExtractor.process(screen, background, frame, image)
              if(mode == Mode.DECLASH) {
                composeScreen(frame, image)
              }
            }
          } else if(mode == Mode.DECLASH && ONLY_BACKGROUND.isEmpty()) {
            composeScreen(frame, toImage(screen, true))
          }
        }
      }

      oldScreen = screen
      frame++
    }
    when(mode) {
      Mode.FIND_PIXELS_TO_SKIP -> {
        for(background in backgrounds) {
          var maxChanged = 0
          if(background.frame < 0) continue
          val area = load(background.frame, MAIN_SCREEN)
          for(addr in 0 until PIXEL_SIZE) {
            if(background.changed!![addr] > maxChanged) {
              maxChanged = background.changed!![addr]
            }
            if(1.0 * background.changed!![addr] / background.total
              >= MIN_BG_CHANGED) {
              area.pixels[addr] = Pixel.ANY
            } else {
              area.pixels[addr] = background.pixels[addr]
            }
          }
          println("Background ${background.name} max changed $maxChanged)" +
              " of ${background.total}")
          val image = toImage(area, false)
          saveImage(image, background.fileName)
        }
      } Mode.COLOR_BACKGROUNDS -> {
        for(background in backgrounds) {
          if(background.frame < 0) continue
          val area = load(background.frame, MAIN_SCREEN)
          val image = toImage(area, true)
          saveImage(image, background.fileName)
        }
      } Mode.EXTRACT_SPRITES -> {
        ImageExtractor.saveImages()
      }
    }

  }

  private fun composeScreen(frame: Int, image: BufferedImage) {
    val i = if(TWO_FRAMES) 1 else 0
    val screenWidth = WHOLE_SCREEN.pixelWidth()
    val screenHeight = WHOLE_SCREEN.pixelHeight()
    val width = screenWidth * 2 + 7
    val height = width * 9 / 16
    val newImage = if(TWO_FRAMES) {
      BufferedImage(width, height, BufferedImage.TYPE_INT_RGB)
    } else {
      BufferedImage(screenWidth, screenHeight,BufferedImage.TYPE_INT_RGB)
    }

    val dy = if(TWO_FRAMES) (height - (SCREEN_HEIGHT shl 3) - 4) / 2 else 0

    val g2d: Graphics2D = newImage.createGraphics()
    if(TWO_FRAMES) {
      g2d.color = Color.black
      g2d.fillRect(0, 0, newImage.width, newImage.height)
      g2d.color = Color.white
      g2d.drawRect(0, dy - 2, newImage.width - 1
        , screenHeight + 3)
      g2d.drawRect(screenWidth + 3, dy - 2, 0
        , screenHeight + 3)

      val fm: FontMetrics = g2d.getFontMetrics(g2d.font)
      g2d.drawString("NEW", 130 - fm.stringWidth("NEW") / 2
        , 206 + dy)
      g2d.drawString("ORIGINAL", 130 + 256
          - fm.stringWidth("ORIGINAL") / 2, 206 + dy)
    }

    pasteToImage(newImage, load(frame, STATUS_BAR)
      , (STATUS_BAR.x shl 3) + i * 2, (STATUS_BAR.y shl 3) + dy, true)
    g2d.drawImage(image, (MAIN_SCREEN.x shl 3) + i * 2
      , (MAIN_SCREEN.y shl 3) + dy, null)
    if(TWO_FRAMES) {
      g2d.drawImage(toImage(load(frame, WHOLE_SCREEN), true)
        , screenWidth + 5 * i, dy, null)
    }

    g2d.dispose()
    saveImage(newImage, frame)
  }

  // conversion and saving
  private fun toImage(area: Area, colored: Boolean
                      , bgData: Array<Pixel>? = null): BufferedImage {
    val image = BufferedImage(area.area.pixelWidth(), area.area.pixelHeight()
      , BufferedImage.TYPE_INT_RGB)
    pasteToImage(image, area, 0, 0, colored, bgData)
    return image
  }

  private fun pasteToImage(image: BufferedImage, area: Area, x0: Int, y0: Int
                           , colored: Boolean, bgData: Array<Pixel>? = null) {
    val width = area.area.pixelWidth()
    val height = area.area.pixelHeight()
    for(y in 0 until height) {
      val ySource = y shl 8
      val yAttrSource = (y shr 3) shl 5
      for(x in 0 until width) {
        val attr = area.attrs[yAttrSource or (x shr 3)]
        val addr = ySource or x
        val value = area.pixels[addr]
        image.setRGB(x + x0, y + y0,
          if(BLACK_AND_WHITE) {
            if(value == Pixel.ON) white else black
          } else if(bgData != null && value != bgData[addr]
            && bgData[addr] != Pixel.ANY && value != Pixel.ANY) {
            magenta
          } else if(colored) {
            if(value == Pixel.ON) {
              color[attr and 0b1111]
            } else {
              color[attr shr 4 and 0b1111]
            }
          } else if(value == Pixel.ANY) {
            magenta
          } else if(value == Pixel.ON) {
            white
          } else {
            black
          }
        )
      }
    }
  }

  @Throws(IOException::class)
  fun saveImage(image: BufferedImage, fileNumber: Int) {
    saveImage(image, format(fileNumber) + ".png")
  }

  @Throws(IOException::class)
  fun saveImage(image: BufferedImage, fileName: String) {
    val outputFile = File(OUT_DIR + fileName)
    ImageIO.write(x3(image), "png", outputFile)
  }
}